src/sig_stfl/xmss/external/xmss_core_fast.c

```diff
@@ -0,0 +1,1099 @@
+        if (states[i].auth == NULL) {
+            ret = -1;
+            goto cleanup;
+        }
+        memcpy(sm, states[i].auth, params->tree_height*params->n);
+        sm += params->tree_height*params->n;
+        *smlen += params->tree_height*params->n;
+    }
+
+    updates = (params->tree_height - params->bds_k) >> 1;
+
+    set_tree_addr(addr, (idx_tree + 1));
+    // mandatory update for NEXT_0 (does not count towards h-k/2) if NEXT_0 exists
+    if ((1 + idx_tree) * (1ULL << params->tree_height) + idx_leaf < (1ULL << (unsigned long long) params->full_height)) {
+        bds_state_update(params, &states[params->d], sk_seed, pub_seed, addr);
+    }
+
+    for (i = 0; i < params->d; i++) {
+        // check if we're not at the end of a tree
+        if (! (((idx + 1) & ((1ULL << ((i+1)*params->tree_height)) - 1)) == 0)) {
+            idx_leaf = (uint32_t)((idx >> (params->tree_height * i)) & ((1 << params->tree_height)-1));
+            idx_tree = (idx >> (params->tree_height * (i+1)));
+            set_layer_addr(addr, i);
+            set_tree_addr(addr, (uint32_t)idx_tree);
+            if (i == (unsigned int) (needswap_upto + 1)) {
+                bds_round(params, &states[i], idx_leaf, sk_seed, pub_seed, addr);
+            }
+            updates = bds_treehash_update(params, &states[i], updates, sk_seed, pub_seed, addr);
+            set_tree_addr(addr, (idx_tree + 1));
+            // if a NEXT-tree exists for this level;
+            if ((1 + idx_tree) * (1ULL << params->tree_height) + idx_leaf < (1ULL << (params->full_height - params->tree_height * i))) {
+                if (i > 0 && updates > 0 && states[params->d + i].next_leaf < (1ULL << params->full_height)) {
+                    bds_state_update(params, &states[params->d + i], sk_seed, pub_seed, addr);
+                    updates--;
+                }
+            }
+        }
+        else if (idx < (1ULL << params->full_height) - 1) {
+            deep_state_swap(params, &states[params->d + i], &states[i]);
+
+            set_layer_addr(ots_addr, (uint32_t)(i+1));
+            set_tree_addr(ots_addr, ((idx + 1) >> ((i+2) * params->tree_height)));
+            set_ots_addr(ots_addr, (((idx >> ((i+1) * params->tree_height)) + 1) & ((1ULL << params->tree_height)-1)));
+
+            wots_sign(params, wots_sigs + i*params->wots_sig_bytes, states[i].stack, sk_seed, pub_seed, ots_addr);
+
+            states[params->d + i].stackoffset = 0;
+            states[params->d + i].next_leaf = 0;
+
+            updates--; // WOTS-signing counts as one update
+            needswap_upto = (int)i;
+            for (j = 0; j < params->tree_height-params->bds_k; j++) {
+                states[i].treehash[j].completed = 1;
+            }
+        }
+    }
+
+    xmssmt_serialize_state(params, sk, states);
+
+cleanup:
+    OQS_MEM_insecure_free(treehash);
+    OQS_MEM_insecure_free(states);
+    OQS_MEM_insecure_free(tmp);
+    OQS_MEM_insecure_free(m_with_prefix);
```

Comment:
Are we confident that none of these contain sensitive data and thus don't need `OQS_MEM_secure_free`? Similarly elsewhere in this file.

Reply:
These variables are stored in stack originally. 
My rewrite convert the variable length stack to heap. I'll check if any of those are sensitive.

Reply:
Yup, these are sensitive states and should be zeroed on heap. Fix in #1790

Reply:
I double-check all places in `xmss/` that uses  `insecure_free`, they are either producing public key, or hash a piece of messages, or verify the message, these state are not sensitive.